hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/launcher/ContainerLaunch.java

```diff
@@ -1382,73 +1414,64 @@ public void setExitOnFailure() {
         return Collections.emptySet();
       }
       final Set<String> deps = new HashSet<>();
-      // env/whitelistedEnv dump values inside double quotes
-      boolean inDoubleQuotes = true;
-      char c;
-      int i = 0;
-      final int len = envVal.length();
-      while (i < len) {
-        c = envVal.charAt(i);
-        if (c == '"') {
-          inDoubleQuotes = !inDoubleQuotes;
-        } else if (c == '\'' && !inDoubleQuotes) {
-          i++;
-          // eat until closing simple quote
-          while (i < len) {
-            c = envVal.charAt(i);
-            if (c == '\\') {
-              i++;
-            }
-            if (c == '\'') {
-              break;
-            }
-            i++;
-          }
-        } else if (c == '\\') {
-          i++;
-        } else if (c == '$') {
-          i++;
-          if (i >= len) {
-            break;
+      fillDeps(deps, envVal);
+      return deps;
+    }
+
+    private void fillDeps(final Set<String> deps, final String input) {
+      Matcher matcher = mainPattern.matcher(input);
+      while (matcher.find()) {
+        // simple $variableName extraction
+        extractVariableNameToDeps(deps, matcher);
+
+        String doubleQuoted = matcher.group(DOUBLE_QUOTED);
+        if (doubleQuoted != null) {
+          // env/whitelistedEnv dump values inside double quotes
+          // single quotes inside double quotes are ignored
+          doubleQuoted =
+              singleQuotePattern.matcher(doubleQuoted).replaceAll("");
+          // the remaining is processed again
+          fillDeps(deps, doubleQuoted);
+        }
+
+        // https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter
```

Comment:
Done, but checkstyle will fail on this line.

The @SuppressWarnings("checkstyle:linelength") could have been used, but it would have a bigger scope than just one line.